### PR TITLE
refactor: introduce `QueryContext` trait

### PR DESCRIPTION
**Pull request by arendjr**
I initially tried to introduce additional traits from the bottom up, similar to how I had done for `Context`. Unfortunately, when I tried to move onto `Binding` I got stuck. This is because some places have to remain hard-coded to `NodeWithSource` or `MarzanoContext` for now, while others had to become abstract, and compile issues popped up at the edges.

So I bit the bullet, and switched to a top-down approach: I added a generic to `Pattern` itself. I notice I even resolved a `TODO` in the code, so I think the idea is quite uncontroversial. It was kinda unavoidable it would become another big PR though :)

### CODE REVIEW

File: crates/core/src/pattern/or.rs [hunk 1]
```diff
@@ -1,46 +1,51 @@
...
             for p in self.patterns.iter() {
                 // filter out pattern which cannot match because of a mismatched node type
-                if let (Some(binding_node), Pattern::ASTNode(node_pattern)) =
+                if let (Some(binding_node), Pattern::AstNode(node_pattern)) =
                     (binding_vector.last().and_then(Binding::as_node), p)
                 {
-                    if node_pattern.sort != binding_node.node.kind_id() {
+                    // Safety: This is safe as long as `MarzanoProblemContext` is the
+                    // only implementation of `ProblemContext`.
+                    if !node_pattern.matches_kind_of(unsafe { transmute(&binding_node) }) {
```
Comment (arendjr) [inline crates/core/src/pattern/or.rs:48]: This bears some attention. This is another edge where I couldn't convert the concrete `NodeWithSource` into an abstract `AstNode`. The proper fix would be to make `Binding` a trait, and then `ResolvedPattern` would be pulled in as either needing another generic or itself becoming a trait too, and the PR would've become a lot bigger yet. So it's my intention to fix this in the next PR instead.

File: crates/core/src/context.rs [hunk 2]
```diff
@@ -10,25 +11,34 @@ use crate::{
...
 };
 use anyhow::Result;
+use grit_util::AstNode;
 use marzano_language::target_language::TargetLanguage;
 use marzano_util::analysis_logs::AnalysisLogs;
 
-pub trait Context {
-    fn pattern_definitions(&self) -> &[PatternDefinition];
+/// Contains various kinds of context about the problem being executed.
+pub trait ProblemContext: Clone + std::fmt::Debug + Sized {
```
Comment (morgante) [inline crates/core/src/context.rs:19]: I'm trying to phase out the "Problem" nomenclature. Let's call this `QueryContext`.

**Review by morgante — COMMENTED** [review]
Looks reasonable and I think a generic is probably necessary.

Let's call it a "query" instead of a "problem" though.

File: crates/core/src/problem.rs [hunk 3]
```diff
@@ -652,3 +654,12 @@ impl MatchRanges {
         }
     }
 }
+
+#[derive(Clone, Debug)]
+pub struct MarzanoProblemContext;
```
Comment (morgante) [inline crates/core/src/problem.rs:659]:
```suggestion
pub struct MarzanoQueryContext;
```

**Review by morgante — CHANGES_REQUESTED** [review]
WASM build is failing.

File: crates/core/src/pattern/or.rs (hunk 1 above)
Comment (morgante) [reply crates/core/src/pattern/or.rs:48]: Thanks, definitely make this a priority.

**Review by morgante — APPROVED** [review]

**Pull request merged**